src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Formatting/DefaultRazorFormattingService.cs

```diff
@@ -63,17 +65,33 @@ internal class DefaultRazorFormattingService : RazorFormattingService
                 throw new ArgumentNullException(nameof(documentSnapshot));
             }
 
-            if (range is null)
-            {
-                throw new ArgumentNullException(nameof(range));
-            }
-
             if (options is null)
             {
                 throw new ArgumentNullException(nameof(options));
             }
 
             var codeDocument = await documentSnapshot.GetGeneratedOutputAsync();
+
+            // Range formatting happens on every paste, and if there are Razor diagnostics in the file
+            // that can make some very bad results. eg, given:
+            //
+            // |
+            // @code {
+            // }
+            //
+            // When pasting "<button" at the | the HTML formatter will bring the "@code" onto the same
+            // line as "<button" because as far as it's concerned, its an attribute.
+            //
+            // To defeat that, we simply don't do range formatting if there are diagnostics.
+
+            // Despite what it looks like, codeDocument.GetCSharpDocument().Diagnostics is actually the
+            // Razor diagnostics, not the C# diagnostics 🤦‍
+            if (range is not null &&
+                codeDocument.GetCSharpDocument().Diagnostics.Any(d => range.OverlapsWith(d.Span.AsVSRange(codeDocument.GetSourceText()))))
```

Comment:
Do we also put RazorSyntaxTree diagnostics on here? Wondering if we should have an extension method `codeDocument.GetDiagnostics()` which pulls from C# document, Razor Syntax Tree, IntermediateDocument etc.

Reply:
I'm not against this idea, and I actually thought these diagnostics would come from the razor syntax tree, but thats not what I saw in practice. Do you know if the razor compiler does report diagnostics on the tree, and for what? Would be good to have more test data.

Reply:
IIRC the compiler reports diagnostics on the syntax tree for Syntax diagnostics, the CSharp document for code gen related errors and IR document for IR related errors. It's been a while though so I could be a 1000% wrong.

Reply:
Yeah, I get the concept, but I don't know what an example of a syntax diagnostic is. I would have thought this test case would have been, as its an unclosed tag, but apparently not. I can do some experimenting I guess.